libraries/AP_Scripting/modules/speedpi.lua

```diff
@@ -0,0 +1,155 @@
+--[[
+
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+   SpeedPI
+   A simple "PI" controller for airspeed. Copied from Andrew Tridgell's original
+   work on the ArduPilot Aerobatics Lua scripts.
+
+    Usage: 
+    1. drop it in the scripts/modules directory
+    2. include in your own script using 
+         local speedpi = requre("speedpi.lua")
+    3. create an instance - you may need to tune the gains
+        local speed_controller = speedpi.speed_controller(0.1, 0.1, 2.5, airspeed_min, airspeed_max)
+    4. call it's update() from your update() with the current airspeed and airspeed error
+        local airspeed_new = speed_controller.update(vehicle_airspeed, desired_airspeed - vehicle_airspeed)
+    5. Set the vehicle airspeed based on the airspeed_new value returned from speedpi    
+
+--]]
+
+local SpeedPI = {}
+
+SpeedPI.SCRIPT_VERSION = "4.6.0-004"
+SpeedPI.SCRIPT_NAME = "Speed PI Controller"
+SpeedPI.SCRIPT_NAME_SHORT = "SpeedPI"
+
+-- constrain a value between limits
+function SpeedPI.constrain(v, vmin, vmax)
+    if v < vmin then
+       v = vmin
+    end
+    if v > vmax then
+       v = vmax
+    end
+    return v
+end
+
+function SpeedPI.PI_controller(kP,kI,iMax,min,max)
+    -- the new instance. You can put public variables inside this self
+    -- declaration if you want to
+    local self = {}
+
+    -- private fields as locals
+    local _kP = kP or 0.0
+    local _kI = kI or 0.0
+    local _iMax = iMax
+    local _min = min
+    local _max = max
+    local _last_t = nil
+    local _I = 0
+    local _P = 0
+    local _total = 0
+    local _counter = 0
+    local _target = 0
+    local _current = 0
+    local nowPI = millis():tofloat() * 0.001
+ 
+    -- update the controller.
+    function self.update(target, current)
+       local now = millis():tofloat() * 0.001
+       if not _last_t then
+          _last_t = now
+       end
+       local dt = now - _last_t
```

Comment:
Do the DT first and then convert to float. That prevent lose of precision  at millis get larger.